### ISSUE TITLE
Code Ready Containers becomes 'Not Reachable' due to etcd crashing and unable to restart

### ISSUE THREAD

**Issue by timroster** [code-ready/crc#1888]
### General information

  * OS: Linux 
  * Hypervisor: KVM
  * Did you run `crc setup` before starting it (Yes/No)? Yes

## CRC version
```bash
# Put the output of `crc version`
CodeReady Containers version: 1.21.0+68a4cdd7
OpenShift version: 4.6.9 (embedded in executable)
```
  
## CRC status
```bash
# Put the output of `crc status`
CRC VM:          Running
OpenShift:       Not Reachable (v4.6.9)
Disk Usage:      25.71GB of 74.6GB (Inside the CRC VM)
Cache Usage:     27.04GB
Cache Directory: /home/crcuser/.crc/cache
```

## CRC config
```bash
# Put the output of `crc config view`
- consent-telemetry                     : no
- cpus                                  : 12
- disk-size                             : 70
- enable-cluster-monitoring             : true
- memory                                : 48000
```

## Host Operating System
```bash
# Put the output of `cat /etc/os-release` in case of Linux
NAME="CentOS Linux"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
```

### Steps to reproduce

  1. Start CRC, leave it running
  2. After 1-7 days (perhaps more... unclear)
  3. CRC OpenShift API stops responding and status shows 'Not Reachable'
  4. Some container workloads (e.g. other pods, services, routes for applications) stay operational
  5. Stopping / restarting `crc` does not recover

### Expected
CRC OpenShift API functions allowing `oc login` and other actions on the cluster against the API:
```
oc login -u developer -p developer
Login successful.

You have one project on this server: "victim"

Using project "victim".
```

### Actual
Unable to login with cli or web console:
```
oc login -u developer -p developer
error: dial tcp 192.168.130.11:6443: connect: connection refused - verify you have provided the correct host and port and that the server is currently running.
```

### Logs

You can start crc with `crc start --log-level debug` to collect logs.
Link to [gist with logs](https://gist.github.com/timroster/59277c5f07cd5698d5b6e6fe17cc6791)

**Comment by timroster** [comment]
pretty much a duplicate of https://github.com/code-ready/crc/issues/1786

copying my notes from that to here:

Possible to recover using `crc stop; crc delete; crc start`

When restarting `crc` w/o using delete the etcd pod is failing at the environment variable checking InitContainer with a log message of (after ssh to crc vm):

```
# cat etcd-crc-fdm75-master-0_openshift-etcd_etcd-ensure-env-vars-d5249245463ac1eadc44c03d0bd8217efd6a24f4f598c8e93c31831e5394be45.log
2020-12-22T00:35:21.651476376+00:00 stderr F Expected node IP to be 192.168.126.11 got 192.168.130.11
```

If crc is restarted with stop then start it will come up ok as this is done before the time when the API becomes unresponsive. This message only appears after a crc status command shows that OpenShift is "Not Reachable".

**Comment by timroster** [comment]
Before restarting to gather debug logs, I also checked on the CRC vm and noted the same issue:

```
# tail -f /var/log/containers/etcd-crc-lf65c-master-0_openshift-etcd_etcd-ensure-env-vars-a9061e40b584245587e27a4efc00c8866c1b8ce500b4f8903e7678294e23e70e.log 
2021-01-26T00:49:15.330307319+00:00 stderr F Expected node IP to be 192.168.126.11 got 192.168.130.11
```

**Comment by timroster** [comment]
@guillaumerose as you commented on the code in the etcd operator in the other issue, I looked a little more at the state of the failing vm configuration and compared that to a working crc instance.

For the failing VM - in the `/etc/kubernetes/manifests/etcd-pod.yaml` file all values that had the IP `192.168.126.11` are now showing `196.168.130.11`

**Comment by timroster** [comment]
In this case, the host machine is an Intel-based VM running on IBM Cloud. I have seen this problem as well on a  I've seen on a native AMD CPU.

One difference between working and failed vm's is the change of IP address in the `/etc/kubernetes/manifests/etcd-pod.yaml` file.

**Comment by timroster** [comment]
Details from `dmidecode` and `lscpu`:  https://gist.github.com/timroster/02b185d97b51fd64c3a77eda1a438e59

**Comment by timroster** [comment]
@gbraad as to running a full OpenShift install to this particular VM type, it's complicated. It may be easier for me to round up an intel-based host to replicate.

**Comment by guillaumerose** [comment]
Thanks @timroster for the issue. There is clearly something wrong with the dummy interface.

**Comment by timroster** [comment]
The dummy interface still is present on the system - and it appears to still be the interface of record for the hostname:

```
/sbin/ip address show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp1s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:fd:fc:07:21:82 brd ff:ff:ff:ff:ff:ff
    inet 192.168.130.11/24 brd 192.168.130.255 scope global dynamic noprefixroute enp1s0
       valid_lft 3051sec preferred_lft 3051sec
    inet6 fe80::50fd:fcff:fe07:2182/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
7: eth10: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 72:ce:aa:19:5d:f5 brd ff:ff:ff:ff:ff:ff
    inet 192.168.126.11/24 brd 192.168.126.255 scope global noprefixroute eth10
       valid_lft forever preferred_lft forever
    inet6 fe80::9bf2:9c50:7e4b:6bef/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
 ...
 ```
 
 ```
 [root@crc-lf65c-master-0 ~]# hostname
crc-lf65c-master-0
[root@crc-lf65c-master-0 ~]# hostname -i
192.168.126.11
[root@crc-lf65c-master-0 ~]# hostname -I
192.168.130.11 192.168.126.11 10.88.0.1 10.116.0.1 
[root@crc-lf65c-master-0 ~]# ping `hostname`
PING crc-lf65c-master-0.crc.testing (192.168.126.11) 56(84) bytes of data.
64 bytes from crc-lf65c-master-0 (192.168.126.11): icmp_seq=1 ttl=64 time=0.123 ms
64 bytes from crc-lf65c-master-0 (192.168.126.11): icmp_seq=2 ttl=64 time=0.037 ms
64 bytes from crc-lf65c-master-0 (192.168.126.11): icmp_seq=3 ttl=64 time=0.042 ms
64 bytes from crc-lf65c-master-0 (192.168.126.11): icmp_seq=4 ttl=64 time=0.037 ms
^C
```

**Comment by cfergeau** [comment]
> For the failing VM - in the `/etc/kubernetes/manifests/etcd-pod.yaml` file all values that had the IP `192.168.126.11` are now showing `196.168.130.11`

So something happened (I'd suspect some openshift component crashed?), which caused `/etc/kubernetes/manifests/etcd-pod.yaml` to be regenerated, which then breaks the cluster because the dummy IP is not used in this manifest.

Does this happen on 'pristine' clusters (`crc delete && crc start` and then wait until it breaks)? Or are you doing more steps before leaving the clusetr for a few days (installing more operators/applications, enabling some operators which are disabled, other modifications to the VM?)

**Comment by cfergeau** [comment]
When your system is failing, can you also check `systemctl status kubelet`?

If I understand your setup correctly, this is a CentOS VM running on IBM cloud on Intel-based hardware,  and this CentOS VM is then used to run crc? Wondering in on the CentOS host, there are some errors in `/var/log/libvirt/qemu/crc.log`?

**Comment by timroster** [comment]
> So something happened (I'd suspect some openshift component crashed?), which caused /etc/kubernetes/manifests/etcd-pod.yaml to be regenerated, which then breaks the cluster because the dummy IP is not used in this manifest.

Yes, that appears to be what has happened, simply patching the manifest does not appear to resolve and a short period after patching it is re-written to the value without the dummy IP

> Does this happen on 'pristine' clusters (crc delete && crc start and then wait until it breaks)? Or are you doing more steps before leaving the clusetr for a few days (installing more operators/applications, enabling some operators which are disabled, other modifications to the VM?)

I've seen it happen at least one time where I have only done a `crc delete && crc start` and just let it sit.

> Once you are ssh'ed into the VM, I wonder if journalctl or sudo crictl ps -a/ sudo crictl logs could be useful to understand what happened. There might be hints buried in there, but openshift logs are very verbose, and I don't know what to look for.

I might have missed a window on this cycle. Since I've restarted the vm to get the debug logs I don't have journal history from the potential event. Will need to look next time.

**Comment by timroster** [comment]
> When your system is failing, can you also check systemctl status kubelet?

gist with [systemctl status kubelet](https://gist.github.com/timroster/b043cd365b168646a6351ba9a5cb51a8)

The output in this capture shows kubelet trying to reach `etcd` at 192.168.130.11... and localhost, both failing as `etcd` is not running. If I check a working crc instance with netstat, the only connections I see to tcp 2379 are to 192.168.126.11.

> If I understand your setup correctly, this is a CentOS VM running on IBM cloud on Intel-based hardware, and this CentOS VM is then used to run crc? Wondering in on the CentOS host, there are some errors in /var/log/libvirt/qemu/crc.log?

Does not appear to be anything untoward between last successful vm start and the point where I shut down the vm to capture the debug logs:

```
2021-01-21 21:03:36.344+0000: Domain id=1 is tainted: host-cpu
char device redirected to /dev/pts/1 (label charserial0)
2021-01-21T21:03:36.762936Z qemu-kvm: -device cirrus-vga,id=video0,bus=pcie.0,addr=0x1: warning: 'cirrus-vga' is deprecated, please use a different VGA card instead
2021-01-26T01:10:55.151492Z qemu-kvm: terminating on signal 15 from pid 154321 (/usr/sbin/libvirtd)
2021-01-26 01:10:55.796+0000: shutting down, reason=shutdown
2021-01-26 01:12:45.031+0000: starting up libvirt version: 6.0.0, package: 28.module_el8.3.0+555+a55c8938 (CentOS Buildsys <bugs@centos.org>, 2020-11-04-01:04:00, ), qemu version: 4.2.0qemu-kvm-4.2.0-34.module_el8.3.0+613+9ec9f184.1, kernel: 4.18.0-193.14.2.el8_2.x86_64, hostname: tor-crc-schematics-instance
```

gist with [full file contents](https://gist.github.com/timroster/3b13811397f98786dded4f4726fadf38)

**Comment by cfergeau** [comment]
The kubelet is still started with `kubelet --node-ip=192.168.126.11` even though its logs show it's trying to access 192.168.130.11

**Comment by timroster** [comment]
One correction related to kubelet. I was looking further and realized that in fact the kubelet logs shown by `systemctl` in the gist are for a sub-process, specifically `kube-apiserver` . When this problem is occurring it looks like the configuration resources in `/etc/kubernetes/manifests` and `/etc/kubernetes/static-pod-resources` for both `etcd` and `kube-apiserver` have changed the target IP from 192.168.126.11 to 192.168.130.11 for the endpoint for `etcd`

**Comment by guillaumerose** [comment]
Very interesting! I started a cluster and look for 192.168.130.11. I don't see it in /etc/kubernetes nor in any etcd keys/values. It must be something updated by an operator.

**Comment by guillaumerose** [comment]
Can you check the content of etcd operator and apiserver operator ? Also look at stopped containers for these containers, there is maybe a clue inside.

**Comment by timroster** [comment]
I appear to have (at least temporarily) recovered the cluster. Several manual things were attempted, as I was experimenting with the pod configmaps in `/etc/kubernetes/static-pod-resources` . This directory appears to hold a history of static pod definitions with multiple versions. I had removed the latest versions, and then updated the `etcd-pod.yaml` values in `/etc/kubernetes/manifests` and then used `crc stop; crc start` without any success - in fact the `etcd-pod.yaml` contents had reverted.

Most recently, I ssh'd to the vm and then updated in place the `/etc/kubernetes/static-pod-resources/etcd-pod-3/configmaps/etcd-pod/pod.yaml` file switching 192.168.130.11 back to 192.168.126.11, followed by fixing IP in `/etc/kubernetes/manifests/etcd-pod.yaml`. Immediately after this I performed `systemctl restart kubelet` .  This time - the kubelet appears to have read the new static pod definitions and the cluster began to recover.

```
$ crc status
CRC VM:          Running
OpenShift:       Degraded (v4.6.9)
Disk Usage:      27.79GB of 74.6GB (Inside the CRC VM)
Cache Usage:     27.04GB
Cache Directory: /home/crcuser/.crc/cache
```

I will monitor, and if it goes out again, I'll see if only the edits to the two resource files for the `etcd` static pod followed by a `systemctl restart kubelet` are sufficient to recover.

**Comment by guillaumerose** [comment]
I think I possibly found the issue. We define 2 times the same parameter node-ip in /etc/systemd/system/kubelet.service

```
ExecStart=/usr/bin/hyperkube \
    kubelet \
      --node-ip=192.168.126.11 \
      --config=/etc/kubernetes/kubelet.conf \
      --bootstrap-kubeconfig=/etc/kubernetes/kubeconfig \
      --kubeconfig=/var/lib/kubelet/kubeconfig \
      --container-runtime=remote \
      --container-runtime-endpoint=/var/run/crio/crio.sock \
      --runtime-cgroups=/system.slice/crio.service \
      --node-labels=node-role.kubernetes.io/master,node.openshift.io/os_id=${ID} \
      --node-ip=${KUBELET_NODE_IP} \
```

I don't know where this KUBELET_NODE_IP env variable is coming but it is definitely suspect!

**Comment by timroster** [comment]
Not sure if that is directly related. But something at some point while the cluster is running seems to be causing the static pod resource for `etcd` to be switched from 192.168.126.11 to 192.168.130.11. And perhaps at the same time, or a later time, when kubelet restarts this pod, the conflict seems to keep it from starting (env-vars initContainer), causing the api server to start failing.

**Comment by guillaumerose** [comment]
I experimented a bit and found out that with cobra, only the the last parameter is accepted. It makes node-ip is always empty in crc.

kubelet code is then doing a lookup to get the IP. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L225

Then, it is compared to the list of all interfaces to see if it's valid (`net.InterfaceAddrs()`). Can you give me the output of `ifconfig` and `route -n` ? 
If it doesn't match, it picks the interface of the default gateway.

I guess we fall in the last case.

**Comment by timroster** [comment]
From the system (now recovered) output is rather long so captured in [ifconfig and route -n gist](https://gist.github.com/timroster/d7966bb2b62b2c6a62b22575148d3e6d)

These don't look too different from any other crc system.

**Comment by guillaumerose** [comment]
Can you try with this bundle and see if it's better? https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/code-ready_snc/315/pull-ci-code-ready-snc-master-e2e-snc/1354454524017774592/artifacts/e2e-snc/gather-snc/

**Comment by timroster** [comment]
> Can you try with this bundle and see if it's better?

Just to confirm steps, use `crc setup` as usual, then replace the qcow2 file with the one from the bundle before `crc start`?

**Comment by timroster** [comment]
figured it out:

```
crc start -p pull-secret
...
INFO Loading bundle: crc_libvirt_4.6.9.crcbundle ... 
INFO Verifying bundle crc_libvirt_4.6.9.crcbundle ... 
INFO Creating CodeReady Containers VM for OpenShift 4.7.0-0.nightly-2021-01-27-110023... 
```

**Comment by gbraad** [comment]
> Thanks @timroster for the issue. There is clearly something wrong with the dummy interface.

https://github.com/code-ready/crc/issues/1786#issuecomment-753557751

Reported that a while back, but wasn't sure about this. Glad this new issue got rid of some noise and moved forward; often several opinions/comments in a single issue make it harder to follow.

**Comment by gbraad** [comment]
> figured it out:
> 
> ```
> INFO Creating CodeReady Containers VM for OpenShift 4.7.0-0.nightly-2021-01-27-110023... 
> ```
> 
> I will keep this instance running and monitor status. May take a little while to see any clear signs one way or another.

OpenShift 4.7 nightly had not been marked as stable last time I checked. Reported this as yellow on the status report for CRC. @guillaumerose ?

You use a different bundle with the `-b` option on start.

**Comment by guillaumerose** [comment]
Yes not the best bundle to test sorry. I need to prepare the snc 4.6 branch for the next release, there will be a better bundle available then.

**Comment by timroster** [comment]
@guillaumerose fwiw - the crc instance that I "recovered" and subject of this issue has failed again, whereas the other instance that I started with the bundle you recommended is still running ok.

@cfergeau looking at relatively recent output from [sudo crictl ps -a](https://gist.github.com/timroster/1e8753a89e76ed70e4701dbfe81d1db2) seems to look like there was some processes kicking off in the cluster 8 and 12 hours before the command was run, and there are a broad range of containers that are cycling due to `etcd` no longer available. Based on `journalctl` timestamps, the failure appeared to be about 8 hours ago.

output from `journalctl` was quite voluminous, but I was able to find the [minute](https://gist.github.com/timroster/82303c5981c5857962ddaa19261ced97) where things appeared to go from ok to bad by searching on `Readiness probe for "etcd-crc`. Critical event appears to be happening here (times UTC):

```
Jan 30 06:39:08 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:08.722255  657366 secret.go:207] Received secret openshift-cluster-version/default-token-b6lgm containing (4) pieces of data, 16604 total bytes
Jan 30 06:39:08 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:08.722087  657366 configmap.go:188] Setting up volume config for pod f867fc1c-6ece-4518-ae47-366636ee052d at /var/lib/kubelet/pods/f867fc1c-6ece-4518-ae47-366636ee052d/volumes/kubernetes.io~configmap/config
Jan 30 06:39:08 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:08.722965  657366 configmap.go:212] Received configMap openshift-cluster-machine-approver/machine-approver-config containing (0) pieces of data, 0 total bytes
Jan 30 06:39:08 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:08.721078  657366 secret.go:183] Setting up volume serving-cert for pod 77e7a34e-2384-4fe6-bf94-4f708f705b9b at /var/lib/kubelet/pods/77e7a34e-2384-4fe6-bf94-4f708f705b9b/volumes/kubernetes.io~secret/serving-cert
Jan 30 06:39:08 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:08.723326  657366 secret.go:207] Received secret openshift-cluster-version/cluster-version-operator-serving-cert containing (2) pieces of data, 4408 total bytes
Jan 30 06:39:08 crc-lf65c-master-0 systemd[1]: var-lib-containers-storage-overlay-f103536cdc9d497779f99fc4b38fc945cad4b6a914b9ff47c623d77ff967dd79-merged.mount: Consumed 0 CPU time
Jan 30 06:39:08 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:08.801030  657366 prober.go:126] Liveness probe for "dns-default-j5mbf_openshift-dns(2713c6c8-bf0b-4b70-86ba-b7f9ee86cb92):dns" succeeded
Jan 30 06:39:08 crc-lf65c-master-0 crio[2947]: time="2021-01-30 06:39:08.810244239Z" level=info msg="Removed container f62a82007e3cbd6d8a9a90e9e6eed0728ca4e8364f071abc0109099d5e62b8c6: openshift-apiserver/apiserver-664d95b7b9-8pqh6/openshift-apiserver-check-endpoints" id=92b30c3d-5c20-4a3e-922b-f40a6a5450de name=/runtime.v1alpha2.RuntimeService/RemoveContainer
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.026067  657366 file.go:201] Reading config file "/etc/kubernetes/manifests/etcd-pod.yaml"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: E0130 06:39:09.028045  657366 file.go:108] Unable to process watch event: can't process config file "/etc/kubernetes/manifests/etcd-pod.yaml": /etc/kubernetes/manifests/etcd-pod.yaml: couldn't parse as pod(Object 'Kind' is missing in 'null'), please check config file
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.028282  657366 file.go:201] Reading config file "/etc/kubernetes/manifests/etcd-pod.yaml"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.036731  657366 kubelet.go:1844] SyncLoop (REMOVE, "file"): "etcd-crc-lf65c-master-0_openshift-etcd(5e4e08ff-7916-4fe5-8388-3f19cd1c0b9f)"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.036895  657366 kubelet_pods.go:1212] marking pod pending termination "5e4e08ff-7916-4fe5-8388-3f19cd1c0b9f"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.037160  657366 kubelet.go:1834] SyncLoop (ADD, "file"): "etcd-crc-lf65c-master-0_openshift-etcd(66d597e3-33b3-44d8-9d5e-296d86fbe8c3)"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.037717  657366 topology_manager.go:233] [topologymanager] Topology Admit Handler
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.037927  657366 kubelet_pods.go:1247] Killing unwanted pod "etcd-crc-lf65c-master-0"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.038167  657366 kuberuntime_container.go:635] Killing container "cri-o://aeb9c6fff966f1cf16dabb76c1455ae29c7fef396eac8ece754af5861c45dc69" with a 30 second grace period
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.038168  657366 kuberuntime_container.go:635] Killing container "cri-o://9948985c0e41c648e3866d1e0c3f7fd8dbf78f8f1185bbf344d7e3369ef7324c" with a 30 second grace period
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.038221  657366 event.go:291] "Event occurred" object="openshift-etcd/etcd-crc-lf65c-master-0" kind="Pod" apiVersion="v1" type="Normal" reason="Killing" message="Stopping container etcdctl"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.078116  657366 event.go:291] "Event occurred" object="openshift-etcd/etcd-crc-lf65c-master-0" kind="Pod" apiVersion="v1" type="Normal" reason="Killing" message="Stopping container etcd"
Jan 30 06:39:09 crc-lf65c-master-0 hyperkube[657366]: I0130 06:39:09.078152  657366 event.go:291] "Event occurred" object="openshift-etcd/etcd-crc-lf65c-master-0" kind="Pod" apiVersion="v1" type="Normal" reason="Killing" message="Stopping container etcd-metrics"
```

After this etcd never fully restarts to pass readiness checks. It looks like there is some corruption of the `/etc/kubernetes/manifests/etcd-pod.yaml` to me from the error, or at least an issue in reading the file. Maybe this causes it to be re-synced, and in the process the IP address gets changed. Sort of speculation...

After a few timeouts in health checks, about 3 minutes later the kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver goes offline and never recovers:

```
Jan 30 06:41:25 crc-lf65c-master-0 hyperkube[657366]: I0130 06:41:25.268581  657366 prober.go:126] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" succeeded
Jan 30 06:41:35 crc-lf65c-master-0 hyperkube[657366]: I0130 06:41:35.259032  657366 prober.go:126] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" succeeded
Jan 30 06:41:45 crc-lf65c-master-0 hyperkube[657366]: I0130 06:41:45.259149  657366 prober.go:126] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" succeeded
Jan 30 06:41:55 crc-lf65c-master-0 hyperkube[657366]: I0130 06:41:55.275737  657366 prober.go:126] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" succeeded
Jan 30 06:42:05 crc-lf65c-master-0 hyperkube[657366]: I0130 06:42:05.267048  657366 prober.go:126] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" succeeded
Jan 30 06:42:15 crc-lf65c-master-0 hyperkube[657366]: I0130 06:42:15.264939  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): HTTP probe failed with statuscode: 500
Jan 30 06:42:25 crc-lf65c-master-0 hyperkube[657366]: I0130 06:42:25.265320  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): HTTP probe failed with statuscode: 500
Jan 30 06:42:35 crc-lf65c-master-0 hyperkube[657366]: I0130 06:42:35.261177  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): HTTP probe failed with statuscode: 500
Jan 30 06:42:50 crc-lf65c-master-0 hyperkube[657366]: I0130 06:42:50.409196  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver" failed (failure): Get "https://192.168.126.11:6443/healthz": dial tcp 192.168.126.11:6443: connect: connection refused
Jan 30 06:43:05 crc-lf65c-master-0 hyperkube[657366]: I0130 06:43:05.245746  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): Get "https://192.168.126.11:17697/healthz": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Jan 30 06:43:10 crc-lf65c-master-0 hyperkube[657366]: I0130 06:43:10.408849  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver" failed (failure): Get "https://192.168.126.11:6443/healthz": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Jan 30 06:43:15 crc-lf65c-master-0 hyperkube[657366]: I0130 06:43:15.246660  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): Get "https://192.168.126.11:17697/healthz": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Jan 30 06:43:20 crc-lf65c-master-0 hyperkube[657366]: I0130 06:43:20.409739  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver" failed (failure): Get "https://192.168.126.11:6443/healthz": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Jan 30 06:43:25 crc-lf65c-master-0 hyperkube[657366]: I0130 06:43:25.247370  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): Get "https://192.168.126.11:17697/healthz": net/http: TLS handshake timeout
Jan 30 06:43:45 crc-lf65c-master-0 hyperkube[657366]: I0130 06:43:45.245856  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): Get "https://192.168.126.11:17697/healthz": dial tcp 192.168.126.11:17697: connect: connection refused
Jan 30 06:44:05 crc-lf65c-master-0 hyperkube[657366]: I0130 06:44:05.245706  657366 prober.go:117] Liveness probe for "kube-apiserver-crc-lf65c-master-0_openshift-kube-apiserver(9c9a7be9-44dd-46f8-96d9-064e3bdaa0de):kube-apiserver-check-endpoints" failed (failure): Get "https://192.168.126.11:17697/healthz": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

**Comment by timroster** [comment]
Also - it was possible to recover the cluster by using the following:

```
# sed -i 's/192.168.130.11/192.168.126.11/g' /etc/kubernetes/manifests/etcd-pod.yaml
# sed -i 's/192.168.130.11/192.168.126.11/g' /etc/kubernetes/static-pod-resources/etcd-pod-6/configmaps/etcd-pod/pod.yaml
# sed -i 's/192.168.130.11/192.168.126.11/g' /etc/kubernetes/static-pod-resources/etcd-pod-6/etcd-pod.yaml
# systemctl restart kubelet
```

Note that the latest generation `/etc/kubernetes/static-pod-resources/etcd-pod-##` should be targeted, the timestamp of this directory matched when the failure appeared - apparently a new static generation is being created.

**Comment by guillaumerose** [comment]
Every 24 hours kubelet node certs are rotated. The cluster machine approver approves it. This is what we see here.
Maybe the rotation triggers a reload of kubelet and it re-reads all options (and network interfaces?).

**Comment by timroster** [comment]
It doesn't seem to happen every 24 hours however. I've seen crc run for a few days before this has happened on some systems. I've never seen it run for an indefinite time without it happening, but maybe I'm unlucky.

I wrote a [short script](https://gist.github.com/timroster/7eb49e14495347f5d5d5a7cbd5d959fd) that can be run from the VM that seems to repair the failures I've seen on KVM with Intel (nested) and AMD (host) virtualization.

**Comment by guillaumerose** [comment]
Hi @timroster ,
The release is almost out. Can you try with http://mirror.openshift.com/pub/openshift-v4/clients/crc/1.22.0/ ? 
Thanks.

**Comment by timroster** [comment]
@guillaumerose - thanks for heads up. Spun up new vm and installed this version. I see a single command line argument for `--node-ip=192.168.126.11` for **kubelet** in the process table. Will let it run and also subject it to some use.

**Comment by timroster** [comment]
6 days running w/o a crash or other problem with the CRC api server availability

**Comment by guillaumerose** [comment]
Great!

**Comment by timroster** [comment]
@guillaumerose I've not seen 1.22 fail with this type of problem on two different systems and with >7 days running time. closing and thanks for the investigation and fixes.
